ospec.js

```diff
@@ -278,116 +547,176 @@ else window.o = m()
+	// 		})
+	// 	spy.args = []
+	// 	spy.calls = []
+	// 	spy.callCount = 0
+	// 	return spy
+	// }
+
+	var spyFactoryCache = Object.create(null)
+
+	function makeSpyFactory(name, length) {
+		if (spyFactoryCache[name] == null) spyFactoryCache[name] = []
+		var args = Array.apply(null, {length: length}).map(
+			function(_, i) {return "_" + i}
+		).join(", ");
+		var code =
+			"'use strict';" +
+			"var spy = (0, function " + name + "(" + args + ") {" +
+			"   return helper(this, [].slice.call(arguments), fn, spy)" +
+			"});" +
+			"return spy"
+
+		return spyFactoryCache[name][length] = new Function("fn", "helper", code)
+	}
+
+	function getOrMakeSpyFactory(name, length) {
+		return spyFactoryCache[name] && spyFactoryCache[name][length] || makeSpyFactory(name, length)
+	}
+
+	function spyHelper(self, args, fn, spy) {
+		spy.this = self
+		spy.args = args
+		spy.calls.push({this: self, args: args})
+		spy.callCount++
+
+		if (fn) return fn.apply(self, args)
+	}
+
+	var supportsFunctionMutations = false;
+	// eslint-disable-next-line no-empty, no-implicit-coercion
+	try {supportsFunctionMutations = !!Object.defineProperties(function(){}, {name: {value: "a"},length: {value: 1}})} catch(_){}
+
+	var supportsEval = false
+	// eslint-disable-next-line no-eval, no-empty
+	try {eval("supportsEval = true")} catch(e){}
+
+	o.spy = function spy(fn) {
+		var name = "", length = 0
+		if (fn) name = fn.name, length = fn.length
+		var spy = (!supportsFunctionMutations && supportsEval)
+			? getOrMakeSpyFactory(name, length)(fn, spyHelper)
+			: function(){return spyHelper(this, [].slice.call(arguments), fn, spy)}
+		if (supportsFunctionMutations) Object.defineProperties(spy, {
+			name: {value: name},
+			length: {value: length}
+		})
+
+		spy.args = []
+		spy.calls = []
+		spy.callCount = 0
+		return spy
 	}
+
+	// Reporter
 	var colorCodes = {
 		red: "31m",
 		red2: "31;1m",
 		green: "32;1m"
 	}
+
+	// console style for terminals
+	// see https://stackoverflow.com/questions/4842424/list-of-ansi-color-escape-sequences
 	function highlight(message, color) {
 		var code = colorCodes[color] || colorCodes.red;
 		return hasProcess ? (process.stdout.isTTY ? "\x1b[" + code + message + "\x1b[0m" : message) : "%c" + message + "%c "
 	}
+
+	// console style for the Browsers
+	// see https://developer.mozilla.org/en-US/docs/Web/API/console#Styling_console_output
 	function cStyle(color, bold) {
 		return hasProcess||!color ? "" : "color:"+color+(bold ? ";font-weight:bold" : "")
 	}
-	function ensureStackTrace(error) {
-		// mandatory to get a stack in IE 10 and 11 (and maybe other envs?)
-		if (error.stack === undefined) try { throw error } catch(e) {return e}
-		else return error
-	}
-	function getStackName(e, exp) {
-		return e.stack && exp.test(e.stack) ? e.stack.match(exp)[1] : null
-	}
 
-	o.report = function (results) {
-		var errCount = 0
+	o.report = function (results, stats) {
+		if (arguments.length === 1) stats = {bailCount: 0, asyncSuccesses: 0}
```

Comment:
Much simpler to check the variable itself, and it keeps the intuition that passing `undefined` is the same as not passing it at all (which is a lot easier to reason about). Plus, we don't do this kind of thing anywhere else.

```suggestion
		if (stats == null) stats = {bailCount: 0, asyncSuccesses: 0}
```